walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,148 @@
+  }
+
+  private boolean isValidParams(Collection<Event> events, long duration) {
+    // TODO: check for events that all events not crossing EOD.
+
+    if (duration > TimeRange.WHOLE_DAY.duration() || duration <= 0) {
+      return false;
+    }
+
+    return true;
+  }
+
+  private ImmutableList<TimeRange> getOccupiedTimeRange(Set<String> attendees, Collection<Event> events) {
+    List<TimeRange> occupiedTimeRange = new ArrayList<>();
+    for (Event event: events) {
+      if (hasIntersection(event.getAttendees(), attendees)) {
+        occupiedTimeRange.add(event.getWhen());
+      }
+    }
+    return ImmutableList.copyOf(occupiedTimeRange);
+  }
+
+  private boolean hasIntersection(Set<String> eventAttendees, Set<String> requestAttendees) {
+    return !Collections.disjoint(eventAttendees, requestAttendees);
+  }
+
+  private ImmutableList<TimeRange> getAvailableTimeRange(ImmutableList<TimeRange> occupiedTimeRange, long duration) {
+    // Assumes 0 < duration <= WHOLE_DAY
+
+    List<TimeRange> availableMeetings = new ArrayList<>();
+
+    if (occupiedTimeRange.isEmpty()) {
+      return ImmutableList.of(TimeRange.WHOLE_DAY);
+    }
+
+    ImmutableList<TimeRange> mergedTimeRanges = mergeTimeRange(occupiedTimeRange);
+
+    int start = TimeRange.START_OF_DAY;
+
+    for (TimeRange timeRange: mergedTimeRanges) {
+      // Interval duration is not enough.
+      if ((timeRange.start() - start) < duration) {
+        start = Math.min(timeRange.end(), TimeRange.END_OF_DAY);
+        continue;
+      }
+
+      TimeRange availableMeeting = TimeRange.fromStartEnd(start, timeRange.start(), /* inclusive= */false);
+      availableMeetings.add(availableMeeting);
+
+      start = Math.min(timeRange.end(), TimeRange.END_OF_DAY);
+    }
+
+    // Deals with the last remaining time slot of a day.
+    if (start < TimeRange.END_OF_DAY && (TimeRange.END_OF_DAY - start) >= duration) {
+      availableMeetings.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, /* inclusive= */true));
+    }
+
+    return ImmutableList.copyOf(availableMeetings);
+  }
+
+  private ImmutableList<TimeRange> mergeTimeRange(ImmutableList<TimeRange> timeRanges) {
+    List<TimeRange> mergedTimeRanges = new ArrayList<>();
+
+    if (timeRanges.isEmpty()) {
+      return ImmutableList.copyOf(mergedTimeRanges);
+    }
+
+    ImmutableList<TimeRange> sortedTimeRanges = ImmutableList.sortedCopyOf(TimeRange.ORDER_BY_START, timeRanges);
+
+    TimeRange last = null;
+    for (TimeRange timeRange: sortedTimeRanges) {
+      // The first TimeRange or no overlapping.
+      if (last == null || !timeRange.overlaps(last)) {
+        mergedTimeRanges.add(timeRange);
+        last = timeRange;
+      } else {
+        TimeRange mergedTimeRange;
+        if (last.contains(timeRange)) {
+          continue;
+        }
+
+        mergedTimeRange = TimeRange.fromStartEnd(last.start(), timeRange.end(), /* inclusive= */false);
+
+        mergedTimeRanges.remove(mergedTimeRanges.size() - 1);
```

Comment:
can you add a comment what this is about

Reply:
Fixed